### PR TITLE
Fix a bug where a gRPC ServerCall is invoked after a stream is closed

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

When a stream is closed by a remote peer, `ArmeriaServerCall` is closed by `res.whenComplete()` listener.
https://github.com/line/armeria/blob/6e24e7817f8be94ab21ddba7879e9f84ca840229/grpc/src/main/java/com/linecorp/armeria/server/grpc/ArmeriaServerCall.java#L173-L182

If a service performs a long running task, `ArmeriaServerCall.sendHeaders()` or
`ArmeriaServerCall.sendMessage()` will be called when a response is written.
Because gRPC-Java `ServerCallStreamObserver` implementation invokes `ServerCall` even if a call is cancelled with some conditions.
https://github.com/grpc/grpc-java/blob/d52b359631ee1ac46f47d72d1015e1dd6ce9312c/stub/src/main/java/io/grpc/stub/ServerCalls.java#L346-L358

Modifications:

- Check `cancelled` flag before validating the closing states.

Result:

- `ArmeriaServerCall` is now closed exatly once.

### CODE REVIEW

**Comment by okue** [comment]
In addition to client close case, doesn't `ctx.setRequestTimeout` close?

https://github.com/line/armeria/blob/armeria-1.2.0/grpc/src/main/java/com/linecorp/armeria/server/grpc/FramedGrpcService.java#L208

**Comment by ikhoon** [comment]
> In addition to the case of client close, isn't there the other case of close caused by `ctx.setRequestTimeout`?
> 
> [`armeria-1.2.0`/grpc/src/main/java/com/linecorp/armeria/server/grpc/FramedGrpcService.java#L208](https://github.com/line/armeria/blob/armeria-1.2.0/grpc/src/main/java/com/linecorp/armeria/server/grpc/FramedGrpcService.java#L208)

I realized that `cancelled` is set true when:
- a call is closed by a remote peer
- a call is closed with non-OK status

Therefore the timeout logic makes `cancelled` true. I need to update comments in the code. 😁

File: grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientTimeoutTest.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    @Test
+    void clientTimeout() throws InterruptedException {
+        final TestServiceBlockingStub client =
+                Clients.newClient(server.httpUri(GrpcSerializationFormats.PROTO),
+                                  TestServiceBlockingStub.class);
+        final StatusRuntimeException exception = catchThrowableOfType(() -> {
+            client.withDeadlineAfter(1000, TimeUnit.MILLISECONDS)
+                  .unaryCall(SimpleRequest.getDefaultInstance());
+        }, StatusRuntimeException.class);
+        assertThat(exception.getStatus().getCode()).isEqualTo(Code.DEADLINE_EXCEEDED);
```
Comment (okue) [inline grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientTimeoutTest.java:97]: If the understanding is correct,

Armeria gRPC client sets timeout when either option `Client.responseTimeout` or `io.grpc.Stub.withDelay` is specified.
After timed out, the client raises DEADLINE_EXCEEDED.

https://github.com/line/armeria/blob/armeria-1.2.0/grpc/src/main/java/com/linecorp/armeria/internal/client/grpc/ArmeriaClientCall.java#L189-L207

Besides, if the either options of them is specified, the client send `grpc-timeout` to the server, and the server also set timeout.
After timed out at the server side, the server returns CANCELLED to the client.

https://github.com/line/armeria/blob/armeria-1.2.0/grpc/src/main/java/com/linecorp/armeria/internal/client/grpc/ArmeriaClientCall.java#L398
https://github.com/line/armeria/blob/armeria-1.2.0/grpc/src/main/java/com/linecorp/armeria/server/grpc/FramedGrpcService.java#L190

So, I think there are two possibilities of the error code raised by the client.

Comment (ikhoon) [reply grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientTimeoutTest.java:97]: > If the understanding is correct,

Yes, exactly.

a gRPC client actively raises `DEADLINE_EXCEEDED` or `CANCELLED` by retruned status code.

Comment (ikhoon) [reply grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientTimeoutTest.java:99]: JFYI (I guess you already know this code), the following mapppings converter an exception to status code. 
https://github.com/line/armeria/blob/1b6d85fa0493f39c6d0b5b5084fe8440bf521223/grpc/src/main/java/com/linecorp/armeria/internal/common/grpc/GrpcStatus.java#L78-L112

Comment (okue) [reply grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientTimeoutTest.java:99]: In this test case, does the client never raise `CANCELLED`?

Comment (ikhoon) [reply grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientTimeoutTest.java:99]: Good point. It is possbile. I think it depends on Operating System scheduling. 
- A client sends a remaining timeout to a server. When the server schedules the timeout, network latency is not considered. Therefore, in most case, client will close a stream if a deadline exceeded.
 
- However, a client CPU is busy and hard to get CPU cycles to schedule a timeout for the deadline, a server timeout scheduler will be triggered and close a stream before the client deadline exceeded.

Comment (ikhoon) [reply grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientTimeoutTest.java:99]: Eventually to resolve this problem, as @trustin said in [slack channel](https://line-armeria.slack.com/archives/C1NGPBUH2/p1604287343245400?thread_ts=1604070308.229600&cid=C1NGPBUH2), we need to unify the timeout.

**Review by minwoox — APPROVED** [review]
Thanks a lot!

**Review by trustin — APPROVED** [review]
Thanks, @ikhoon!

**Comment by ikhoon** [comment]
Fixed #3153

**Pull request merged**